enumerable.rb

```diff
@@ -1,87 +1,163 @@
+# frozen_string_literal: true
+
+# rubocop: disable Metrics/ModuleLength
+# rubocop: disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+
+# module Enumerable
 module Enumerable
   def my_each
-    for i in self do
-        self.class == Array ? yield(i) : yield(i[0], i[1])
+    return to_enum(:my_each) unless block_given?
+
+    i = 0
+    while i < to_a.length
+      instance_of?(Array) ? yield(to_a[i]) : yield(to_a[i][0], to_a[i][1])
+      i += 1
     end
+    self
   end
 
   def my_each_with_index
-    for i in self do
-      self.class == Array ? yield(i, self.index(i)) : yield([i[0], i[1]], self.keys.index(i[0]))
+    return to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    while i < to_a.length
+      instance_of?(Array) ? yield(to_a[i], i) : yield([to_a[i][0], to_a[i][1]], i)
+      i += 1
     end
+    self
   end
 
   def my_select
+    return to_enum(:my_select) unless block_given?
+
     new_arr = []
     new_hash = {}
-    self.my_each do |k,v|
-        if self.class == Array
-            new_arr << k if yield(k) == true
-        else
-            new_hash.store(k,v) if yield(k,v) == true
-        end
+    my_each do |k, v|
+      if instance_of?(Array)
+        new_arr << k if yield(k) == true
+      elsif yield(k, v) == true
+        new_hash.store(k, v)
+      end
     end
     new_arr.empty? ? new_hash : new_arr
   end
 
   def my_all?(*args)
     if block_given?
-      self.my_each do |e|
+      my_each do |e|
         return false if yield(e) == false
       end
     end
-    if !block_given? 
-      if self.index(nil)
-        return !self.index(nil) ? true : false
-      end  
-      if !args.empty?
-        if ( args[0].class != Regexp) && ( args[0].class != Class)
-          return  false 
-        end
-        self.my_each do |e|
-          if ( args[0].class == Regexp) &&(args[0].match(e.to_s)) == nil
-            return false
-          end
-          if ( args[0].class == Class) && (e.is_a? args[0]) == false 
-            return false
-          end
+    unless block_given?
+      if index(nil)
+        return !index(nil) ? true : false
+      end
```

Comment:
- [ ] when no block or argument is given returns false when one of the collection members is false or nil
```
  p [1, false, 'hi', []].all?  # false
  p [1, false, 'hi', []].my_all? # true
```

- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```
  p [3,3,3,3].all?(3) # true
  p [3,3,3,3].my_all?(3) # False
```